capstone/src/test/java/com/google/sps/RecommenderAlgorithmTest.java

```diff
@@ -0,0 +1,422 @@
+
+        EmbeddedEntity userTagMap = TEST_USER_SMITH.getCachedInterestVector();
+        EmbeddedEntity otherTagMap = TEST_USER_JOHN.getCachedInterestVector();
+
+        Double distances = RecommenderAlgorithm.distanceBetween(userTagMap, otherTagMap);
+        Assert.assertEquals(distances, 0.7653668647301795, epsilon);
+    }
+
+    @Test
+    public void calculateDistanceTwoUsersTwoSimilarRoomsMultipleRoomsTest() { 
+        TEST_ROOM_RUN.addTag(Tag.FITNESS);
+        TEST_ROOM_RUN.addFollower(TEST_USER_SMITH);
+        TEST_ROOM_RUN.addFollower(TEST_USER_JOHN);
+        DatabaseService.save(TEST_ROOM_RUN.getRoomEntity());
+
+        TEST_ROOM_COOK.addTag(Tag.COOKING);
+        TEST_ROOM_COOK.addFollower(TEST_USER_SMITH);
+        TEST_ROOM_COOK.addFollower(TEST_USER_JOHN);
+        DatabaseService.save(TEST_ROOM_COOK.getRoomEntity());
+
+        TEST_ROOM_POETRY.addTag(Tag.LITERATURE);
+        TEST_ROOM_POETRY.addFollower(TEST_USER_SMITH);
+        DatabaseService.save(TEST_ROOM_POETRY.getRoomEntity());
+
+        EmbeddedEntity userTagMap = TEST_USER_SMITH.getCachedInterestVector();
+        EmbeddedEntity otherTagMap = TEST_USER_JOHN.getCachedInterestVector();
+
+        Double distances = RecommenderAlgorithm.distanceBetween(userTagMap, otherTagMap);
+        Assert.assertEquals(distances, 0.6058108930553725, epsilon);
+    }
+
+    @Test
+    public void calculateShortestDistanceMultipleUsersMultipleRoomsTest() { 
+        TEST_ROOM_RUN.addTag(Tag.FITNESS);
+        TEST_ROOM_RUN.addFollower(TEST_USER_SMITH);
+        TEST_ROOM_RUN.addFollower(TEST_USER_JOHN);
+        TEST_ROOM_RUN.addFollower(TEST_USER_SAM);
+        DatabaseService.save(TEST_ROOM_RUN.getRoomEntity());
+
+        TEST_ROOM_COOK.addTag(Tag.COOKING);
+        TEST_ROOM_COOK.addFollower(TEST_USER_SMITH);
+        TEST_ROOM_COOK.addFollower(TEST_USER_JOHN);
+        DatabaseService.save(TEST_ROOM_COOK.getRoomEntity());
+
+        TEST_ROOM_POETRY.addTag(Tag.LITERATURE);
+        TEST_ROOM_POETRY.addFollower(TEST_USER_SMITH);
+        TEST_ROOM_POETRY.addFollower(TEST_USER_JOHN);
+        TEST_ROOM_POETRY.addFollower(TEST_USER_SAM);
+        DatabaseService.save(TEST_ROOM_POETRY.getRoomEntity());
+
+        TEST_ROOM_COOK_TWO.addTag(Tag.COOKING);
+        TEST_ROOM_COOK_TWO.addFollower(TEST_USER_SMITH);
+        DatabaseService.save(TEST_ROOM_COOK_TWO.getRoomEntity());
+
+        EmbeddedEntity smithTagMap = TEST_USER_SMITH.getCachedInterestVector();
+        EmbeddedEntity johnTagMap = TEST_USER_JOHN.getCachedInterestVector();
+        EmbeddedEntity samTagMap = TEST_USER_SAM.getCachedInterestVector();
+
+        Double smithJohnDistance = RecommenderAlgorithm.distanceBetween(smithTagMap, johnTagMap);
+        Double smithSamDistance = RecommenderAlgorithm.distanceBetween(smithTagMap, samTagMap);
+
+        Assert.assertEquals(smithJohnDistance, 0.33820395745152554, epsilon);
+    }
+
+    /* ------------------------------------------------------------ */
```

Comment:
All the test above do not need to use DatabaseService.save(). For the tests below, we really should be testing for findSimilarRooms() and change the function implementation so that it doesn't talk to the database. Because of time, I decided to leave the test as is and in a future branch (we all know this won't happen xD) we can remove the DatabaseService.save() and further separate application and intregration logic.

Reply:
Also most of these tests are credited to Kyle. I decided to copy and paste from PR#25 instead of merging since it seemed easier so it doesn't have his commits. All I changed was the tag name and remove the normalize embedded entity.